pkg/grpc/pool.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpc
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	clocklib "github.com/benbjohnson/clock"
+	"google.golang.org/grpc"
+
+	"github.com/dapr/kit/ptr"
+)
+
+// Real-time clock (wrapper around time.Time) to allow mocking
+var clock = clocklib.New()
+
+// Maximum number of concurrent streams in a single gRPC connection
+// This is the default value used by gRPC servers and clients
+const grpcMaxConcurrentStreams = 100
```

Comment:
Is this configurable?

Reply:
Sort of, but on the servers (so each sidecar can control the max concurrent streams incoming only). It's something configured at the HTTP/2 level actually: https://httpwg.org/specs/rfc7540.html#rfc.section.5.1.2 In Go, the default is "at least 100" (https://pkg.go.dev/golang.org/x/net/http2#Server)